src/Compiler/Microsoft.AspNetCore.Razor.Language/src/Syntax/SyntaxListBuilderExtensions.cs

```diff
@@ -11,20 +9,41 @@ public static SyntaxList<SyntaxNode> ToList(this SyntaxListBuilder builder)
     {
         if (builder == null || builder.Count == 0)
         {
-            return default(SyntaxList<SyntaxNode>);
+            return default;
+        }
+
+        return new SyntaxList<SyntaxNode>(builder.ToListNode().AssumeNotNull().CreateRed());
+    }
+
+    public static SyntaxList<SyntaxNode> ToList(this SyntaxListBuilder builder, SyntaxNode parent)
+    {
+        if (builder == null || builder.Count == 0)
+        {
+            return default;
         }
 
-        return new SyntaxList<SyntaxNode>(builder.ToListNode().CreateRed());
+        return new SyntaxList<SyntaxNode>(builder.ToListNode().AssumeNotNull().CreateRed(parent, parent.Position));
     }
 
     public static SyntaxList<TNode> ToList<TNode>(this SyntaxListBuilder builder)
         where TNode : SyntaxNode
     {
         if (builder == null || builder.Count == 0)
         {
-            return new SyntaxList<TNode>();
+            return default;
+        }
+
+        return new SyntaxList<TNode>(builder.ToListNode().AssumeNotNull().CreateRed());
+    }
+
+    public static SyntaxList<TNode> ToList<TNode>(this SyntaxListBuilder builder, SyntaxNode parent)
+        where TNode : SyntaxNode
+    {
+        if (builder == null || builder.Count == 0)
+        {
+            return default;
         }
 
-        return new SyntaxList<TNode>(builder.ToListNode().CreateRed());
+        return new SyntaxList<TNode>(builder.ToListNode().AssumeNotNull().CreateRed(parent, parent.Position));
```

Comment:
Seems like we could use `!` in places that would fail with NullReferenceException immediately anyway.

Reply:
`NullReferenceException` doesn't give as good of a message though. `AssumeNotNull()` is intended to state the programmer's intent (which `!` doesn't really do) and it uses `CallerArgumentExpression` to provide the expression that actually failed in the expression.